resources/table.ml

```diff
@@ -0,0 +1,53 @@
+module Make (V : sig
+    type t
+
+    val safe_release : t -> unit
+
+    val move_out_exn : t -> t
+
+    val with_moved_exn: t -> (t -> 'd) -> 'd
+  end) =
+struct
+  open Xapi_stdext_threads.Threadext
+  type 'a t = ('a, V.t) Hashtbl.t * Mutex.t
+
+  let create n : 'a t = Hashtbl.create n, Mutex.create ()
+
+  let release (t, m) =
+    Mutex.execute m (fun () ->
+      Hashtbl.iter (fun _ v -> V.safe_release v) t)
+
+  let reset (t, m) = release (t, m) ; Hashtbl.reset t
+
+  let copy (t, m) =
+    Mutex.execute m (fun () -> Hashtbl.copy t), Mutex.create ()
+
+  let length (t, m) =
+    Mutex.execute m (fun () -> Hashtbl.length t)
+
+  let move_into (t, m) k v =
+    let v = V.move_out_exn v in
+    Mutex.execute m (fun () -> Hashtbl.add t k v)
```

Comment:
`Hashtbl.add` would shadow an existing entry. Is this correct or would it be safer to use `Hashtbl.replace`?

Reply:
It seems the old code used add, but I agree about replace. Also have to release the old value, so it is a few more lines, see my latest commit.